### PR TITLE
AWS S3 Bucket for FReDT

### PR DESCRIPTION
**Pull request by caseyli92**
DESCRIPTION OF PR:

* Fetch sea level rise data and load it into memory instead of downloading the sea level rise datasets
* Fetch backup REC data from NIWA OpenData instead of loading the manually downloaded backup data file
* Update `river` module to use S3 bucket
* Update `flood_model` module to use S3 bucket
* Update logging format and loggers to exclude

## Developer Checklist
- [x] Make code change
- [ ] Update tests
  - [ ] Update / create new tests
  - [ ] Ensure these tests have the expected behaviour
  - [ ] Test locally and ensure tests are passing
- [x] Update documentation
  - [ ] Readme
  - [x] Docstrings
  - [x] Comments
  - [ ] Wiki

## Reviewer Checklist
- [ ] Check new code for code smells
- [ ] Check new tests
  - [ ] Ensure adequate coverage
  - [ ] Check for code smells within tests
- [ ] Check if documentation needs updating
  - [ ] Readme
  - [ ] Docstrings
  - [ ] Comments
  - [ ] Wiki

### CODE REVIEW

File: .env.template
Comment (LukeParky) [inline .env.template:10]: So AWS S3 is optional right? That sounds like a good plan

**Review by LukeParky — APPROVED** [review]
Looks very good, I like your usage of the S3Manager.
I added some comments that i'd like you to review and then I am happy for it to be merged. Not all of them need to be fixed but have a look at each of them.

Because its a major core feature change make sure it's tested with completely empty database volumes etc and `USE_AWS_S3_BUCKET=False` and before merging

File: .env.template
Comment (caseyli92) [reply .env.template:10]: Yes, optional :)

File: environment.yml
Comment (caseyli92) [reply environment.yml:31]: It is required for loading netcdf file into memory from S3 bucket, used in `s3_connection.py` line 104:
`xr.load_dataset(body_object, engine="h5netcdf")`

File: selected_polygon.geojson
Comment (caseyli92) [reply selected_polygon.geojson:37]: It is still the Kaiapoi area, I've put it back to the original catchment area we were previously using, not sure how it had been changed once.

Comment (LukeParky) [reply selected_polygon.geojson:37]: I made it bigger so that it was easier to preload the whole area and then draw rectangles within it, but let me know if that was causing issues

File: environment.yml
Comment (LukeParky) [reply environment.yml:31]: That sounds good. Could you lock down the version for it as well, for consistent builds.

I looked into the library and it looks like it can be a drop-in replacement for `python-netcdf4` which I thought we used but looking closer we don't use it anymore.

File: src/digitaltwin/s3_connection.py
Comment (caseyli92) [reply src/digitaltwin/s3_connection.py:62]: It's hard to say it depends on what the boto3 functions can take.

File: src/digitaltwin/utils.py
Comment (caseyli92) [reply src/digitaltwin/utils.py:6]: yes, line 56

File: src/dynamic_boundary_conditions/river/river_network_to_from_db.py
Comment (caseyli92) [reply src/dynamic_boundary_conditions/river/river_network_to_from_db.py:196]: Good idea.

File: selected_polygon.geojson
Comment (caseyli92) [reply selected_polygon.geojson:37]: I thought it was done by accident, so I've put it back so that we didn't need to re-generate the HydroDEM for the area. Do you want me to put it back to the bigger area?

File: environment.yml
Comment (caseyli92) [reply environment.yml:31]: locked down h5netcdf version to 1.3.0

File: src/dynamic_boundary_conditions/river/river_network_to_from_db.py
Comment (caseyli92) [reply src/dynamic_boundary_conditions/river/river_network_to_from_db.py:297]: done

File: src/dynamic_boundary_conditions/river/river_network_to_from_db.py
Comment (caseyli92) [reply src/dynamic_boundary_conditions/river/river_network_to_from_db.py:196]: Changed in multiple places.

File: src/digitaltwin/utils.py
Comment (LukeParky) [reply src/digitaltwin/utils.py:6]: Okay cool, I see it. It must have been from a previous PR

File: selected_polygon.geojson
Comment (LukeParky) [reply selected_polygon.geojson:37]: Yes please, I chose a bigger area because the old one was too small, hopefully it doesn't cause too much re-generation

Comment (caseyli92) [reply selected_polygon.geojson:37]: done

File: src/dynamic_boundary_conditions/tide/sea_level_rise_data.py
Comment (caseyli92) [reply src/dynamic_boundary_conditions/tide/sea_level_rise_data.py:127]: After each dataset has been retrieved, loaded, and concatenated, the count would increase by 1. I think when all loaded it will always evaluate to True which is the case here, so this `count` variable is probably not even needed, but I thought to keep it just in case, coz we had something similar in the old code. Whats your thoughts?

File: src/dynamic_boundary_conditions/tide/sea_level_rise_data.py
Comment (caseyli92) [reply src/dynamic_boundary_conditions/tide/sea_level_rise_data.py:182]: `prep_slr_geo_data` converts the pandas DataFrame to GeoDataFrame, I have moved `prep_slr_geo_data` to the end of the `fetch_slr_data_from_takiwa` function now, so `fetch_slr_data_from_takiwa` returns a GeoDataFrame now.

File: src/dynamic_boundary_conditions/tide/sea_level_rise_data.py
Comment (caseyli92) [reply src/dynamic_boundary_conditions/tide/sea_level_rise_data.py:127]: Now I think we should just remove it. Since it is not downloading the files anymore.

Comment (caseyli92) [reply src/dynamic_boundary_conditions/tide/sea_level_rise_data.py:127]: `len(slr_data)` returns the number of rows in the `slr_data` dataframe.

File: src/dynamic_boundary_conditions/tide/sea_level_rise_data.py
Comment (caseyli92) [reply src/dynamic_boundary_conditions/tide/sea_level_rise_data.py:123]: I have split the extracting region name bit out into a new function called `extract_region_name`, have a look and let me know what you think or if there are better ways you'd like this function to be split.

File: src/dynamic_boundary_conditions/tide/sea_level_rise_data.py
Comment (LukeParky) [reply src/dynamic_boundary_conditions/tide/sea_level_rise_data.py:111]: The old code was because the downloads did not always complete with selenium. Now since you are reading the data from memory I think it does not need the `count`

Comment (caseyli92) [reply src/dynamic_boundary_conditions/tide/sea_level_rise_data.py]: I have removed `count`

File: src/digitaltwin/s3_connection.py
Comment (caseyli92) [reply src/digitaltwin/s3_connection.py]: done.

File: src/digitaltwin/s3_connection.py
Comment (caseyli92) [reply src/digitaltwin/s3_connection.py:87]: done.

File: src/dynamic_boundary_conditions/river/river_network_to_from_db.py
Comment (caseyli92) [reply src/dynamic_boundary_conditions/river/river_network_to_from_db.py]: done

**Review by LukeParky — APPROVED** [review]
Looks good. Ready to merge after paper branch is merged and tagged

**Comment by LukeParky** [comment]
Hi @caseyli92, do you think this PR should be merged into the Ōtakaro branch instead of main?

**Comment by caseyli92** [comment]
> Hi @caseyli92, do you think this PR should be merged into the Ōtakaro branch instead of main?

Yes, I think that should be the best way to go.

File: environment.yml
Comment (LukeParky) [inline environment.yml:32]:
```suggestion
  - botocore>=1.33.10 # Minimum version that is compatible with python >= 3.10 is botocore>=1.13.0
  - h5netcdf==1.3.0
```
Could you please remove scrapy in this PR, since it's in newzealidar now

File: .env.docker-override
Comment (LukeParky) [inline .env.docker-override:6]: Associated with these changes, could you also ensure they're removed from `.env.template`, `docker-compose.yml,` and `docker-compose-dev.yml`.
Make sure you're up to date with the base branch when you do this.

Comment (caseyli92) [reply .env.docker-override:6]: Done. Can you please review the changes? @LukeParky

Comment (LukeParky) [reply .env.docker-override:6]: Looks like you aren't up to date with `otakaro_digital_twin` branch so I can't confirm it but it looks right to me. Once you merge in `otakaro_digital_twin` I'll do a full review

File: src/dynamic_boundary_conditions/river/river_network_to_from_db.py
Comment (LukeParky) [inline src/dynamic_boundary_conditions/river/river_network_to_from_db.py:304]:
```suggestion
    use_aws_s3_bucket = EnvVariable.USE_AWS_S3_BUCKET
```

**Review by LukeParky — APPROVED** [review]
Changes look good. Fix the linting errors and do a good test run before merging.

File: src/digitaltwin/s3_connection.py
Comment (LukeParky) [inline src/digitaltwin/s3_connection.py:54]: Same comment as in newzealidar:

I am unsure if `boto3.Session` has good error messages but it might be worth doing some explicit validation that `access_key_id` and `secret_access_key` aren't empty. See what happens when they're empty and make a decision on if it's needed.

**Comment by caseyli92** [comment]
Issue with running `newzealidar` in debug mode still needs to be fixed #278

**Comment by LukeParky** [comment]
Closing for now, because it looks like this will not be needed with AWS S3 mounting available.

**Pull request closed without merging**
